include/trieste/source.h

```diff
@@ -71,46 +71,100 @@ namespace trieste
     {
-      // Lines and columns are 0-indexed.
-      auto it = std::lower_bound(lines.begin(), lines.end(), pos);
+      // If we have no lines, contents is an empty string.
+      // Realistically this case will only happen for pos == 0.
+      if (lines.empty())
+      {
+        return {0, pos};
+      }
 
-      auto line = it - lines.begin();
-      auto col = pos;
+      // Find the first line that begins _after_ pos, and then backtrack one
+      // element if we need to. We can't write this directly because "last
+      // element for which condition was false" is not an stdlib primitive.
+      auto it = std::lower_bound(
+        lines.begin(),
+        lines.end(),
+        pos,
+        [](std::pair<size_t, size_t> elem, size_t pos_) {
+          return elem.first <= pos_;
+        });
+      // If we're at the beginning already, or we couldn't find a line after
+      // pos, then our current line should be fine. If lines is constructed
+      // correctly, it is on either the only line or the last line (and pos is
+      // on that line or beyond all lines).
+      if (it == lines.end() || (it != lines.begin() && it->first > pos))
+      {
+        --it;
+      }
 
-      if (it != lines.begin())
-        col -= *(it - 1) + 1;
+      size_t line = std::distance(lines.begin(), it);
+      size_t col = pos - it->first;
 
       return {line, col};
     }
 
     std::pair<size_t, size_t> linepos(size_t line) const
     {
-      // Lines are 0-indexed.
-      if (line > lines.size())
-        return {std::string::npos, 0};
-
-      size_t start = 0;
-      auto end = contents.size();
-
-      if (line > 0)
-        start = lines[line - 1] + 1;
-
-      if (line < lines.size())
-        end = lines[line];
+      // Special case: for out of range lines, index them at the end of our
+      // string and give them length 0. This will cause minimally-ugly
+      // misbehavior if there's an out of range error. Also, this gracefully
+      // handles the technically out of range situation where we ask for line 0
+      // of an empty string, which is what linecol(0) will give its caller in
+      // that case.
+
+      // Change note: this used to return std::string::npos when line was out of
+      // range by more than 1, but callers weren't checking for it so that case
+      // just caused things like .subtr(max_long) in practice. Best return an
+      // empty line with max idx, which will cause blank outputs. Otherwise,
+      // actually assert(line <= lines.size()) to crash here and not half way
+      // down the calling function.
+      if (line >= lines.size())
+      {
+        return {contents.size(), 0};
+      }
 
-      return {start, end - start};
+      return lines[line]; // already in {start, size} format
     }
 
   private:
     void find_lines()
     {
-      // Find the lines.
-      auto pos = contents.find('\n');
+      using namespace std::string_view_literals;
```

Comment:
As mentioned on the other PR (repeating comment here) I think a slight modification makes this a bit cleaner:

```c++
auto is_eol = [](const std::string_view s) -> std::pair<bool, size_t> {
        if(s.size() > 1 && s.substr(0, 2) == "\r\n"){
            return {true, 2};
        }
        
        if(s.empty()){
            return {false, 0};
        }
       
        if(s[0] == '\n'){
            return {true, 1};
        }
        
        return {false, 1};
    };

    size_t cursor = 0;
    size_t line_start = 0;
    std::string_view contents_view(contents);
    while(cursor < contents_view.size()){
        auto [eol, inc] = is_eol(contents_view.substr(cursor));
        if(eol){
            lines.emplace_back(line_start, cursor);
            cursor += inc;
            line_start = cursor;
        }else{
            cursor += inc;
        }
    }

    if(cursor - line_start > 0){
        lines.push_back({line_start, cursor});
    }
```

Reply:
Style question, now I realize it was the `[&]` you had a problem with. Why is that considered bad?

To me, a big strength of local functions is that you don't have to waste time with convoluted state passing when you can just reference local state (the local function has the same or shorter lifetime than the state, so no memory problems).  I use this a lot in Scala.

If there's a cultural difference in C++, I'd like to understand it properly, because almost every non-trivial accumulator-using function I've written while working on this project uses this style. If I knew what was seen as undesirable about it, I could make a broader change.

Reply:
While I would still like your thoughts on the above, I think the idea to use a string_view as an accumulator is good, and I updated my code.

I simplified it a bit from your example, since `substr` is quite robust, and I changed `try_match` to return the already-shifted string_view so that the position increment and the comparison are just next to each other, not inside and outside the function (easier to see that one guards the other). I used an optional ensure the null case, where the string didn't match, also works.

Reply:
(added a code comment for those who haven't just read the substr edge case docs)

Reply:
This has become mostly irrelevant, given subsequent discussion.